helix-core/src/main/java/org/apache/helix/participant/HelixCustomCodeRunner.java

```diff
@@ -130,14 +147,23 @@ public void start() throws Exception {
 
     StateMachineEngine stateMach = _manager.getStateMachineEngine();
     stateMach.registerStateModelFactory(LEADER_STANDBY, _stateModelFty, _resourceName);
-    HelixZkClient zkClient = null;
+    RealmAwareZkClient zkClient = null;
     try {
       // manually add ideal state for participant leader using LeaderStandby
       // model
-      HelixZkClient.ZkClientConfig clientConfig = new HelixZkClient.ZkClientConfig();
-      clientConfig.setZkSerializer(new ZNRecordSerializer());
-      zkClient = SharedZkClientFactory.getInstance()
-          .buildZkClient(new HelixZkClient.ZkConnectionConfig(_zkAddr), clientConfig);
+      if (Boolean.getBoolean(SystemPropertyKeys.MULTI_ZK_ENABLED) || _zkAddr == null) {
```

Comment:
if zkaddress is null and we are initialize it with FederatedZKClient, do we have problem? For example, we cannot get the routing key and cause any problems?

Reply:
The whole point of FederatedZkClient is not using a particular ZkAddress, but rather rely on RealmAwareConnectionConfig to establish a multi-zk connection, so this makes sense. The sharding key information is included in the connection config.